File: challenger-backend/src/main/java/com/kameo/challenger/odb/TaskProgressODB.java
package com.kameo.challenger.odb;

import com.kameo.challenger.odb.api.IIdentity;
import lombok.Data;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Date;

@Data
@Entity
@Table(indexes = {
    @Index(unique = true, columnList = "task_id, progressTime")
})
public class TaskProgressODB implements IIdentity {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private long id;

    @NotNull
    Date progressTime;

    boolean done;


    @NotNull
    @ManyToOne
    TaskODB task;

    public TaskODB getTask() {
        return task;
    }

    public void setTask(TaskODB task) {
        this.task = task;
    }

    public Date getProgressTime() {
        return progressTime;
    }

    public void setProgressTime(Date progressTime) {
        this.progressTime = progressTime;
    }

    public boolean isDone() {
        return done;
    }

    public void setDone(boolean done) {
        this.done = done;
    }


    @Override
    public boolean isNew() {
        return getId()<0;
    }
}

File: challenger-backend/src/main/java/com/kameo/challenger/domain/tasks/db/TaskProgressODB.kt
package com.kameo.challenger.domain.tasks.db

import com.kameo.challenger.odb.api.IIdentity
import java.time.LocalDate
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.Id
import javax.persistence.ManyToOne
import javax.validation.constraints.NotNull

@Entity
data class TaskProgressODB(@Id
                           @GeneratedValue(strategy = javax.persistence.GenerationType.AUTO)
                           override val id: Long = 0) : IIdentity {



    lateinit var progressTime: LocalDate

    var done: Boolean = false

    @NotNull
    @ManyToOne
    lateinit var task: TaskODB

    constructor(task: TaskODB, progressTime: LocalDate, done: Boolean) : this(0) {
        this.task = task
        this.progressTime = progressTime
        this.done = done
    }
}